istioctl/cmd/add-to-mesh.go

```diff
@@ -0,0 +1,204 @@
+			if err != nil {
+				return err
+			}
+			matchingDeployments, err := findDeploymentsForSvc(client, ns, args[0])
+			if err != nil {
+				return err
+			}
+			if len(matchingDeployments) == 0 {
+				fmt.Fprintf(writer, "No deployments found for service %s.%s", args[0], ns)
+				return nil
+			}
+			return injectSideCarIntoDeployment(client, matchingDeployments, sidecarTemplate, valuesConfig,
+				args[0], ns, meshConfig, writer)
+		},
+	}
+	return cmd
+}
+func setupParameters(sidecarTemplate, valuesConfig *string) (*meshconfig.MeshConfig, error) {
+	var meshConfig *meshconfig.MeshConfig
+	var err error
+	if meshConfigFile != "" {
+		if meshConfig, err = istiocmd.ReadMeshConfig(meshConfigFile); err != nil {
+			return nil, err
+		}
+	} else {
+		if meshConfig, err = getMeshConfigFromConfigMap(kubeconfig); err != nil {
+			return nil, err
+		}
+	}
+	if injectConfigFile != "" {
+		injectionConfig, err := ioutil.ReadFile(injectConfigFile) // nolint: vetshadow
+		if err != nil {
+			return nil, err
+		}
+		var injectConfig inject.Config
+		if err := yaml.Unmarshal(injectionConfig, &injectConfig); err != nil {
+			return nil, multierr.Append(fmt.Errorf("loading --injectConfigFile"), err)
+		}
+		*sidecarTemplate = injectConfig.Template
+	} else if *sidecarTemplate, err = getInjectConfigFromConfigMap(kubeconfig); err != nil {
+		return nil, err
+	}
+	if valuesFile != "" {
+		valuesConfigBytes, err := ioutil.ReadFile(valuesFile) // nolint: vetshadow
+		if err != nil {
+			return nil, err
+		}
+		*valuesConfig = string(valuesConfigBytes)
+	} else if *valuesConfig, err = getValuesFromConfigMap(kubeconfig); err != nil {
+		return nil, err
+	}
+	return meshConfig, err
+}
+
+func injectSideCarIntoDeployment(client kubernetes.Interface, deps []appsv1.Deployment, sidecarTemplate, valuesConfig,
+	svcName, svcNamespace string, meshConfig *meshconfig.MeshConfig, writer io.Writer) error {
+	var errs error
+	for _, dep := range deps {
+		log.Debugf("updating deployment %s.%s with Istio sidecar injected",
+			dep.Name, dep.Namespace)
+		newDep, err := inject.IntoObject(sidecarTemplate, valuesConfig, meshConfig, &dep)
+		if err != nil {
+			errs = multierr.Append(fmt.Errorf("failed to update deployment %s.%s for service %s.%s due to %v",
+				dep.Name, dep.Namespace, svcName, svcNamespace, err), errs)
+			continue
+		}
+		res, b := newDep.(*appsv1.Deployment)
+		if !b {
+			errs = multierr.Append(fmt.Errorf("failed to update deployment %s.%s for service %s.%s",
+				dep.Name, dep.Namespace, svcName, svcNamespace), errs)
+			continue
+		}
+		if _, err := client.AppsV1().Deployments(svcNamespace).Update(res); err != nil {
+			errs = multierr.Append(fmt.Errorf("failed to update deployment %s.%s for service %s.%s due to %v",
+				dep.Name, dep.Namespace, svcName, svcNamespace, err), errs)
+			continue
+
+		}
+		if _, err = client.AppsV1().Deployments(svcNamespace).UpdateStatus(res); err != nil {
+			errs = multierr.Append(err, errs)
```

Comment:
I tried to use add-to-mesh while the pods were still starting.  The message was

```
Error: Operation cannot be fulfilled on deployments.apps "details-v1": the object has been modified; please apply your changes to the latest version and try again
```

This is confusing.  I think the message came from this line but I am not sure.  Use multierr to indicate what failed here.

Reply:
fixed ```Error: failed to update deployment guestbook-v2.default for service guestbook.default due to Operation cannot be fulfilled on deployments.apps "guestbook-v2": the object has been modified; please apply your changes to the latest version and try again
```

Reply:
fixed